service/r2/store/kv/store.go

```diff
@@ -366,28 +373,59 @@ func (s *store) DeleteRollupRule(
 func (s *store) FetchRollupRuleHistory(namespaceID, rollupRuleID string) ([]*models.RollupRuleView, error) {
 	rs, err := s.ruleStore.ReadRuleSet(namespaceID)
 	if err != nil {
-		return nil, s.handleUpstreamError(err)
+		return nil, handleUpstreamError(err)
 	}
 
 	rrs, err := rs.RollupRules()
 	if err != nil {
-		return nil, s.handleUpstreamError(err)
+		return nil, handleUpstreamError(err)
 	}
 
 	hist, exists := rrs[rollupRuleID]
 	if !exists {
-		return nil, s.rollupRuleNotFoundError(namespaceID, rollupRuleID)
+		return nil, rollupRuleNotFoundError(namespaceID, rollupRuleID)
 	}
 	return hist, nil
 }
 
 func (s *store) Close() { s.ruleStore.Close() }
 
-func (s *store) newUpdateMeta(uOpts r2.UpdateOptions) rules.UpdateMetadata {
+func (s *store) UpdateRuleSet(
+	rsChanges changes.RuleSetChanges,
+	version int,
+	uOpts r2store.UpdateOptions,
+) (*models.RuleSet, error) {
+	rs, err := s.FetchRuleSet(rsChanges.Namespace)
+	if err != nil {
+		return nil, handleUpstreamError(err)
+	}
+	if version != rs.Version() {
```

Comment:
This needs to be added here, despite using set and check in KV. This is because the user will pull in the rule set, apply the changes, and then send up the changes. Before we apply the changes locally to send to KV we must check for a version mismatch in case there has been after the user has sent up the proposed changes.

Reply:
Do we need to explicitly check this here? If there are changes after the user has sent up the proposed changes, the check and set will fail at the last step anyway no?

Reply:
This is not the case.  Because changes are applied to an in-memory dataset then committed to KV, at some point m3ctl will have to apply the changes before writing back to KV. So imagine the case where a user sends up some changes and while that request is inflight some new version of rules are landed. The users request will then make it to m3ctl and the updated ruleset will be pulled out so we can apply the changes onto the ruleset to write it back. If the user's changes are applied on top of the changes that just came out of KV there could be conflicts which will be blindly written over, and then sent to KV which will see the changes as valid because the version numbers match up.

Reply:
Hm maybe I'm missing something here, in my mind the typical use case is:
```
1. users fetch latest ruleset, say version = v1 at this point
2. users make changes
3. users send changes alongside version v1
4. m3ctl commits changes by doing a checkandset against version v1.
```
If any thing changes in kv after step 1, that should cause a check and set failure no?

Reply:
well the KV api just allows for setting ruleset. So at some point we need to take ruleset and apply the changes to in application logic before we can commit back to KV. So we must fetch from KV to get a ruleset, apply the changes, and then write it back. But if the version in KV is different from the version the user was looking at when we created the alert, when we apply those changes in application logic we could blow away changes that happen after the user sent the request for changes. For an itemized view it looks more like this.

```
1. users fetch latest ruleset, say version = v1 at this point
2. users make changes
3. users send changes alongside version v1
4. m3ctl fetches changes and applies the changes to create a new ruleset (version 2).
5. m3ctl commits the new ruleset version 2 on top of version 1.
```
If in between stages 3 and 4 a different users updates what is in KV we dont want to try to apply the changes on top of the new version of the ruleset.